protocol-select/README.md

```diff
@@ -0,0 +1,541 @@
+  client (depending on the details of the cryptographic handshake protocol). In
+  that case, the client will be able to immediately open a stream after
+  completing the cryptographic handshake. In addition the protocol supports
+  zero-round-trip optimistic stream protocol negotiation when proposing a single
+  protocol.
+
+- **Data schema**
+
+  The **[Multistream Select]** protocol is defined as a plaintext protocol
+  with no strict schema definition, making both implementation and protocol
+  evolution time consuming and error-prone. See [rust-libp2p/1795] showcasing
+  complexity for implementors and [specs/196] to showcase difficulty evolving
+  protocol.
+
+  The **Protocol Select** protocol will use a binary data format defined in a
+  machine parseable schema language allowing protocol evolution at the schema
+  level.
+
+- **Bandwidth**
+
+  **Multistream Select** is not as bandwidth efficient as it could be. For
+  example negotiating a protocol requires sending the protocol name back and
+  forth. For human readability protocol names are usually long strings (e.g.
+  `/ipfs/kad/1.0.0`).
+
+  **Protocol Select** will include the option to improve bandwidth efficiency
+  e.g. around protocol names in the future. While _Protocol Select_ will not
+  solve this in the first iteration, the protocol is designed with this
+  optimization in mind, and allows for a smooth upgrade in a future iteration.
+
+## High-Level Overview
+
+### Basic Flow
+
+Both endpoints, client and server, send a list of supported protocols. Whether
+an endpoint sends its list before or after it has received the remote's list
+depends on the context and is detailed below. Nodes SHOULD order the list by
+preference. Once an endpoint receives a list from a remote, the protocol to be
+used on the connection or stream is determined by intersecting ones own and the
+remote list, as follows:
+
+1. All protocols that aren't supported by both endpoints are removed from the
+   clients' list of protocols.
+
+2. The protocol chosen is the first protocol of the client's list.
+
+If there is no overlap between the two lists, the two endpoints can not
+communicate and thus both endpoints MUST close the connection or stream.
+
+### Secure Channel Selection
+
+Conversely to [Multistream Select], secure channel protocols are not dynamically
+negotiated in-band. Instead, they are announced upfront in the peer multiaddrs
+(**TODO**: add link to multiaddr spec). This way, implementations can jump
+straight into a cryptographic handshake, thus curtailing the possibility of
+packet-inspection-based censorship and dynamic downgrade attacks.
+
+Given that there is no in-band security protocol negotiation, nodes have to
+listen on different ports for each offered security protocol. As an example a
+node supporting both [Noise] and [TLS] over TCP will need to listen on two TCP
+ports e.g. `/ip6/2001:DB8::/tcp/9090/noise` and `/ip6/2001:DB8::/tcp/443/tls`.
```

Comment:
have we given any thought on how this affect relay connection establishment?

Reply:
Or will that stay in multistream-select?

Reply:
Thanks for taking a look! Can you expand on the problem you are foreseeing with the Multiaddress extension and relay connection establishment?

Reply:
we'd have to advertise `..../p2p-circuit/tls/p2p/...` or else we won't be able to negotiate a security protocol.

Reply:
I think @vyzo is right. Good catch!
Do you see any reason why `..../p2p-circuit/tls/p2p/...` wouldn't be a good idea?

Reply:
In principle it's fine, but I am concerned that a lot of code that handles relay addresses is going to break. So we'd need to be careful with that.

Reply:
Actually... It's even more weird. Currently, `/p2p/QmRelay/p2p-circuit/ip4/1.2.3.4/tcp/1234/p2p/QmTarget` tells the _relay_ to connect to the target at `/ip4/1.2.3.4/tcp/1234`.

In this spec, this would be: `/p2p/QmRelay/p2p-circuit/ip4/1.2.3.4/tcp/1234/tls/p2p/QmTarget`.

However, that _still_ doesn't specify how the local peer should connect to the target. That is, how should we secure the relayed stream? I'm not sure where that would be specified...

Really, we _want_ something like `/p2p/QmRelay/p2p-circuit/{/ip4/1.2.3.4/tcp/1234/tls/,....}/tls/p2p/QmTarget` (where `....` could be additional transport addresses). But the current multiaddr spec doesn't help here.

So, I don't really have a _right_ answer here.

Reply:
We're going to need to redefine _something_ here. My gut reaction is to say that no, the part after `/p2p-circuit` is _not_ treated as an address to give to the relay. But that makes active relaying a bit... tricky.

Reply:
well yeah, and as much as we don't like active relay it is still a legitimate use case for some scenarios.

Reply:
_The below is based on a discussion with @marten-seemann. While we both agree that it ain't elegant, it might either still hold as a viable solution or inspire a more elegant one_.

Taxonomy:

- `A`: The destination node.
- `B`: The source node.
- `R`: The relay node.
- outer connection: The connection from `R` to `A`.
- relayed (inner) connection: The relayed (inner) connection from `B` to `A`.

Problems we need to solve:

1. `B` and `A` need to know which security protocol to use to upgrade the relayed (inner) connection.
2. In the case of active relaying, `R` needs to know which security protocol to use to upgrade the outer connection to the destination.
3. `A` should be able to offer different security protocols for the relayed (inner) connection. `A` can use different layer 4 ports to demultiplex security protocols for the outer connection. `A` has no demultiplexing mechanism for the security protocols used for the relayed (inner) connection. Thus if `A` advertises different multiaddresses with different security protocols for the relayed (inner) connection, it has no way to determine which one to use on incoming relayed (inner) connections.

We propose two mechanisms to solve the issues above:

1. We have to somehow embed both outer and inner security protocol in the multiaddr. What we could do is introduce another separator, e.g. `p2p-circuit-inner` (name yet to be determined).
    - **Solving (1)**: A multiaddr for a passive relayed connection could then look like: `/ip4/../tcp/../tls/p2p/QmRelay/p2p-circuit/p2p/QmA/p2p-circuit-inner/noise`, with `noise` being used to secure the relayed (inner) connection.
    - **Solving (2)**: A multiaddr for an active relayed connection could then look like: `/ip4/../tcp/../tls/p2p/QmRelay/p2p-circuit/ip6/::1/tls/p2p/QmA/p2p-circuit-inner/noise`, with `tls` being used to secure the outer connection and `noise` being used to secure the relayed (inner) connection.
2. **Solving (3)**: In order for `A` to be able to advertise different multiaddresses with different security protocols for the relayed (inner) connection, while still being able to choose the right security protocol on incoming relayed (inner) connections, we need to include the security protocol for the relayed (inner) connection in the circuit relay v2 CONNECT message from `B` to `A`.

Reply:
Given the lack of alternatives, I suggest we integrate the approach above into this specification and https://github.com/libp2p/specs/pull/353. Just to avoid confusion, this should not be interpreted as a final commitment to the suggestion above. Any objections?

Reply:
https://github.com/libp2p/specs/pull/359/ adds the above suggestions to https://github.com/libp2p/specs/pull/353.

Reply:
The above suggestion is now included in https://github.com/libp2p/specs/pull/353. I am resolving this conversation. Please post any follow-ups on https://github.com/libp2p/specs/pull/353.